Add a version strings generator and cache

Konamiman opened the pull request:
_This is the first of a series of pull request to implement output caching for the WooCommerce REST API endpoints. The next one is https://github.com/woocommerce/woocommerce/pull/61798_

_NOTE: The functionality implemented here was initially introduced in https://github.com/woocommerce/woocommerce/pull/61414_

### Changes proposed in this Pull Request:

This pull requests adds the `VersionStringGenerator` class, a WP cache-based version string generation and caching engine. The idea is simple: given a string that represents the id of something, generate and store a random UUID-formatted version so that the next invocation for the same id will return the same version (except if it has expired in the cache); and provide a mechanism for explicitly regenerating the version for a given id to a different value. This is useful for handling cache invalidations when cached objects change.

The class provides the following public methods, each accepting an id:

- `get_version`: If a version is stored for a given entity, return it; otherwise create one and return it. Optional: pass `false` as a second argument, then if no version is stored for that id `null` is returned.
- `generate_version`: Modifies the stored version for a given id, or creates a new one if none exists.
- `delete_version`: Deletes the version stored for a given entity, if it exists.
- `should_use`: Returns true if a persistent object cache is available, false otherwise.

Version keys are cached for one hour by default, but this can be changed via the `woocommerce_version_string_ttl` filter.

### How to test the changes in this Pull Request:

There are comprehensive unit tests for this engine, but it can also be tested using the command line. A prerequisite for that: a persistent object cache installed in your WordPress instance. If you are testing locally you can use [this simple file-based drop-in plugin](https://gist.github.com/Konamiman/96aeacc7933b5b89cb735c46f02adf89).

Run `wp shell`, then execute the following commands in sequence:

1. Get the instance of the class:

```php
$vsg = wc_get_container()->get( Automattic\WooCommerce\Internal\Caches\VersionStringGenerator::class );
```

2. Run the following command, you'll get a random UID:

```php
echo $vsg->get_version( "thing_123" );
```

3. Repeat the same command, you should get the *same* UID.

4. Repeat but using `"another_thing_456"`. You should get a different UID. Verify that repeating step 2 returns the first UID again.

5. Now run the following, you should get an UID that's *different* from the one in step 2:

```php
echo $vsg->generate_version( "thing_123" );
```

6. Repeat step 2, you should get the same UID as in step 5.

7. Run the following command and repeat step 2 twice, you should get a *different* UID (but the same one in both cases):

```php
echo $vsg->delete_version( "thing_123" ) ? "DELETED" : "FAILED";
```

8. Run the following command, you should get "YES" if a global cache is enabled or "NO" otherwise:

```php
echo "Should use: " . ($vsg->should_use() ? "YES" : "NO");
```

9. Run the following command, you should get `null` since the version doesn't exist and we're not generating a new one:

```php
var_dump($vsg->get_version( "nonexistent_id", false ));
```

10. To test custom TTLs, run the following command twice in quick succession, you should get the *same* UID. Wait 5 seconds and run it again, you should get a *different* UID:

```php
add_filter( "woocommerce_version_string_generator_ttl", fn($ttl, $id) => 5, 10, 2 ); 
```

Then repeat step 2 twice in quick succession, you should get the *same* UID. Wait 5 seconds and run it again, you should get a *different* UID.

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

Konamiman committed:
Add EntityVersionsCache class

Introduces a generic mechanism for caching versions of mutable entities,
used for cache invalidation when entities change.

Features:
- Track entity versions by type and ID
- Get/modify/forget entity versions
- Store versions in transients
- Enabled by default only when external object cache is configured
- Filterable TTL and enable/disable via hooks
- Fire actions on version changes

This provides the foundation for REST API response caching.

Konamiman committed:
Add unit tests for EntityVersionsCache

Comprehensive test coverage for EntityVersionsCache including:
- Entity version get/modify/forget operations
- Cache enable/disable behavior
- TTL handling and expiration
- Action hooks firing
- Transient storage and retrieval
- Cache refresh on version retrieval

Konamiman committed:
Add changelog file

Konamiman committed:
Improve the docblock of is_enabled and the woocommerce_enable_entity_versions_cache filter

Konamiman committed:
Add a tearDown method to unit tests

Konamiman committed:
Add arguments and filter return value validations.

joshuatf commented on plugins/woocommerce/src/Internal/Caches/EntityVersionsCache.php:
```diff
@@ -0,0 +1,215 @@
+ *
+ * Provides a generic mechanism for caching versions of mutable entities.
+ */
+class EntityVersionsCache {
```
Just out of curiosity, what do you think the benefits of doing this in a concrete class vs an abstraction where we create individual entity caches is?  E.g., `ProductVersionsCache extends EntityVersionsCache`?

The latter may give us a little bit more flexibility.

ObliviousHarmony replied:
[Your entity invalidation interface PR](https://github.com/woocommerce/woocommerce/pull/61613) could pair nicely with that inheritance scheme. Rather than having an invalidation interface/class, the `ProductVersionsCache` would register all of those hooks and bust the cache for products.

Konamiman replied:
I think composition works best here, so a `ProductVersionsCache` class (possibly implementing `CacheInvalidatorInterface`) would take `EntityVersionsCache` from the container and use it. This provides better separation of concerns (the concept of invalidating products vs the actual act of invalidating it), also this allows concrete implementations like `ProductVersionsCache` to do additional functionality (writing logs, firing another hooks, accessing additional caches that aren't so specific to the REST API...) without polluting the original entity versions cache.

joshuatf commented on plugins/woocommerce/src/Internal/Caches/EntityVersionsCache.php:
```diff
@@ -0,0 +1,215 @@
+		/**
+		 * Filter whether to enable the entity versions cache.
+		 *
+		 * By default, the entity versions cache will be enabled only if an external
```
I notice this text about depending on an object cache, but then we use transients below.

Konamiman replied:
The comment is correct: the class will report to be enabled (that is, `is_enabled` will return true) only if an object cache is present. That's orthogonal to the fact that transients are used (but not really, also see "Note about the usage of transients" in the PR description) for the actual storage. 

Maybe a better name for this method could be `should_use`.

joshuatf commented on plugins/woocommerce/src/Internal/Caches/EntityVersionsCache.php:
```diff
@@ -0,0 +1,215 @@
+	 * @return void
+	 * @throws \InvalidArgumentException If entity_type or entity_id are invalid.
+	 */
+	private function validate_input( string $entity_type, $entity_id ): void {
```
Is this validation helpful beyond the empty string check?  It seems like we could get this for free by having stronger typing on our function arguments.

Konamiman replied:
For `$entity_id`: the entity id can be an input or a string, but unfortunately we need to support PHP 7.4 so we can't add that kind of type checking in the method signature; hence the explicit check.

For `$entity_type`: afaik there's no way to restrict arguments to non-empty strings via explicit typing of function arguments, but happy to be proved wrong.

joshuatf commented on plugins/woocommerce/src/Internal/Caches/EntityVersionsCache.php:
```diff
@@ -0,0 +1,215 @@
+	 * @return mixed The cached value or false if not found.
+	 */
+	protected function get_cached( string $cache_key ) {
+		return get_transient( $cache_key );
```
I know we brought this up in the last PR and on Slack, but I thought the verdict was to move to use the WP persistent object caching instead.

Environments that need it will typically have it set up and we'll avoid further polluting the DB.

ObliviousHarmony replied:
I agree, especially given that this entire system is disabled without an external cache configured in the first place!

Konamiman replied:
> especially given that this entire system is disabled without an external cache configured in the first place

But it can be selectively enabled via the `woocommerce_enable_entity_versions_cache` filter. That's the whole point of using transients: you won't get db-based transients usage (and thus db polluting) unless you explicitly opt-in for it (because wp's transients API will actually use the WP cache), and this is useful for testing/staging scenarios while being harmless for production scenarions.

Konamiman replied:
I have changed the code to use the WP cache, with hooks to allow using a different storage mechanism if needed.

joshuatf commented on plugins/woocommerce/src/Internal/Caches/EntityVersionsCache.php:
```diff
@@ -0,0 +1,215 @@
+	 * @return string The new entity version.
+	 * @throws \InvalidArgumentException If entity_type or entity_id are invalid.
+	 */
+	public function modify_entity_version( string $entity_type, $entity_id ): string {
```
Nitpicky, but it took me a second to grok what/how this function was modifying.  Maybe `generate_new_entity_version()` to match how it's described?

joshuatf commented on plugins/woocommerce/src/Internal/Caches/EntityVersionsCache.php:
```diff
@@ -0,0 +1,215 @@
+		 *
+		 * @since 10.4.0
+		 */
+		$ttl = apply_filters( 'woocommerce_cached_entity_version_ttl', DAY_IN_SECONDS, $entity_type, $entity_id );
```
Nice!  I like the option for plugins to control this.

joshuatf commented on plugins/woocommerce/src/Internal/Caches/EntityVersionsCache.php:
```diff
@@ -0,0 +1,215 @@
+	 * @return bool True on success, false on failure.
+	 * @throws \InvalidArgumentException If entity_type or entity_id are invalid.
+	 */
+	public function forget_entity_version( string $entity_type, $entity_id ): bool {
```
Also nitpicky, but I'd lean into the semantics of `delete_` or `invalidate_` to match typical semantics around changes to cache.

joshuatf commented on plugins/woocommerce/tests/php/src/Internal/Caches/EntityVersionsCacheTest.php:
```diff
@@ -0,0 +1,637 @@
+	 *
+	 * @param bool $wp_using_ext_object_cache_value Value wp_using_ext_object_cache should return.
+	 */
+	public function test_is_enabled_without_filter( bool $wp_using_ext_object_cache_value ) {
```
This test feels a little superfluous with a lot of overridden methods that are not needed and mostly tests the overridden class works in PHP.

joshuatf commented on plugins/woocommerce/src/Internal/Caches/EntityVersionsCache.php:
```diff
@@ -0,0 +1,215 @@
+		 * @param int        $ttl         Time to live in seconds.
+		 * @param bool       $is_new      Whether this is a new version (true) or a refresh of existing version (false).
+		 */
+		do_action( 'woocommerce_entity_version_cached', $entity_type, $entity_id, $ttl, $is_new );
```
Is this needed by consumers?  If it's not absolute necessary and consumers usually rely on just checking the version, I'd prefer to remove this and keep the API surface area a bit more minimal.

Konamiman replied:
I was thinking of possible subsystems that need to act when an entity gets its version changed, but ok, let's keep things simple and we can re-add this later if needed.

Konamiman replied:
Also I guess `woocommerce_entity_version_cache_deleted` can be removed too.

joshuatf reviewed (commented):
This is looking pretty good to me overall!  Left some nitpicky comments, some clarifying comments, and some options to make the API simpler to reduce long-term maintenance.

ObliviousHarmony reviewed (commented):
Thanks @Konamiman, I really like how this is shaping up and it seems like an effective way to identify changes in entity versions.

After reviewing this pull request, to be honest, I'm was super confused! As I understand it, this is a class for managing the current _version string_ of entities, rather than _versions_ of entities? If so, the name `EntityVersionCache` doesn't feel super clear on its purpose. Other than that, I don't really have any criticisms!

---

Just as a thought, what if we changed the relationship between this class and the REST API functionality? The basic idea of providing a mechanism for invalidating cached data when associated entities are changed is really powerful. If that is the intention, having it so far away from the entities themselves feels like it might be a mistake.

1. **`CacheableEntity`**: An interface to be applied to any entity that can be stored in a cache. This would contain a `get_entity_cache_key()` function that returns the ID to be used when busting the cache.
2. **`EntityCache`**: An abstract class containing the version string management functionality currently implemented in something like `invalidate_version( CacheableEntity $entity )` to bust and update the cached version string. Rather than maintaining an entity type, it will just use the class of the cacheable entity as the type. A protected method like `cache_dependent( $data, $key, $entity_ids )` would then handle caching dependent data along with the related entities that invalidate it.
2. **`RESTEntityCache`**: Inherits from `EntityCache` and implements the REST-specific response caching mechanism. This would basically just be your other PR in class form. You would call into this from the request lifecycle then get cached entities and update the cache when necessary. This would have something like `check_response_cache( $request )` or so and would pull from the cache using one of the protected functions and return a typed response!

The end result is that you get a kind of generic mechanism for caching miscellaneous data alongside entities that will invalidate the cache. That could marry well with REST, AJAX, and anything else we dream up.

Konamiman committed:
Misc refactorings:

- Rename `is_enabled` to `should_use()`
- Rename `woocommerce_enable_entity_versions_cache` to
  `woocommerce_should_use_entity_versions_cache`
- Rename `forget_entity_version` to `delete_entity_version`
- Rename `modify_entity_version` to `regenerate_entity_version`
- Remove the `woocommerce_entity_version_cached` and
  `woocommerce_entity_version_cache_deleted` actions
- Use the WP cache instead of transients
- Add "pre_" filter pattern for all cache operations
  (`woocommerce_pre_entity_versions_cache_get/set/delete`)
  to allow custom storage implementations
- Change `get_cached()` to return `null` instead of `false` on cache miss

Konamiman committed:
Rename EntityVersionsCache to EntityVersionKeysCache

Konamiman committed:
Fix a broken test

Konamiman committed:
Remove usage of legacy_proxy for WP cache operations

The problem is that wp_cache_set has a paramter by reference,
and legacy_proxy does not support that.

Konamiman committed:
Simplify the caching logic: wp cache functions are now used directly.
Also the filters to bypass that have been removed.

Konamiman committed:
Fix linting issues

Konamiman committed:
Merge branch 'trunk' into add-entity-versions-cache

Konamiman commented:
@joshuatf @ObliviousHarmony I have simplified the implementation by hardcoding the calls to wp cache functions and removing the filters that allow to use an alternative storage engine for cache entries. Now manual testing always requires a persistent object cache to be installed (testing instructions updated).

Konamiman commented on plugins/woocommerce/src/Internal/Caches/EntityVersionKeysCache.php:
```diff
@@ -0,0 +1,170 @@
+		$this->should_use = is_null( $this->legacy_proxy ) ? false : ( $this->legacy_proxy->call_function( 'wp_using_ext_object_cache' ) ?? false );
+
+		return $this->should_use;
+	}
```
These woocommerce_pre_entity_version_keys_cache_{get,set,delete} filters have been removed to keep the implementations simple.

Konamiman commented on plugins/woocommerce/src/Internal/Caches/EntityVersionKeysCache.php:
```diff
@@ -0,0 +1,170 @@
+			$this->store_entity_version( $entity_type, $entity_id, $version );
+		}
+		return $version;
+	}
```
These woocommerce_pre_entity_version_keys_cache_{get,set,delete} filters have been removed to keep the implementations simple.

Konamiman committed:
Rename EntityVersionKeysCache to VersionStringsGenerator

Konamiman committed:
- Rename public methods to get_version, regenerate_version, delete_version
- Add a new $generate parameter to get_version
- Change methods to accept single string ID parameter instead of entity_type + entity_id
- Hash IDs with MD5 for cache keys to handle special chars and length limits

joshuatf commented on plugins/woocommerce/src/Internal/Caches/VersionStringsGenerator.php:
```diff
@@ -0,0 +1,183 @@
+ * This is useful for cache invalidation strategies where items change over time.
+ * The standard WordPress cache is used to store the version strings.
+ */
+class VersionStringsGenerator {
```
Minor semantic nitpick: 
```suggestion
class VersionStringGenerator {
```

Since it generates a single string.

joshuatf commented on plugins/woocommerce/src/Internal/Caches/VersionStringsGenerator.php:
```diff
@@ -0,0 +1,183 @@
+	 *
+	 * @return bool
+	 */
+	public function should_use(): bool {
```
I feel this `should_use` is more tied to the REST API cache specifics and not to the version string.

I'd be in favor of removing it from this implementation and adding `should_use` wherever we need to check if it should be used.

Alternatively, we could rename this to `can_use` since it does rely on the object cache.

Konamiman replied:
`can_use` is it.

joshuatf commented on plugins/woocommerce/src/Internal/Caches/VersionStringsGenerator.php:
```diff
@@ -0,0 +1,183 @@
+		 *
+		 * @since 10.4.0
+		 */
+		$ttl = apply_filters( 'woocommerce_version_string_ttl', DAY_IN_SECONDS, $id );
```
Very minor but this name might get confused with the actual "woocommerce_version."

Should we rename to:

```suggestion
		$ttl = apply_filters( 'woocommerce_version_string_generator_ttl', DAY_IN_SECONDS, $id );
```

joshuatf commented on plugins/woocommerce/tests/php/src/Internal/Caches/VersionStringGeneratorTest.php:
```diff
@@ -0,0 +1,339 @@
+	 * @testdox get_version returns the existing version if it exists.
+	 */
+	public function test_get_version_returns_existing() {
+		// Pre-populate cache with a known version.
```
Many of these comments seem a bit superfluous.  Are they helpful or AI generated?

The excess use of comments is typically a code smell IMO unless there is specific context the code is missing that requires said comment.

Konamiman replied:
AI generated, but I manually reviewed the result and already removed comments (Claude tends to be very verbose indeed). I'll do another (manual) pass.

joshuatf reviewed (approved):
Pre-approving as I've only left a few nitpicky comments.  Thanks for taking the time do discuss and break this down.

FWIW, I think @ObliviousHarmony's suggestions also make sense.  I'm fine with this being a simple version string generator (as is currently) or an interface that is more coupled to entities.

It might be good to get some buy-in from @ObliviousHarmony, but this looks good to me and at this point, I think we should get this merged to unblock the next pieces of this.

Konamiman committed:
- Rename class from VersionStringsGenerator to VersionStringGenerator
- Rename method should_use() to can_use()
- Rename hook woocommerce_version_string_ttl to woocommerce_version_string_generator_ttl

Konamiman committed:
Make unit tests less verbose
